### PR TITLE
WIP TypeScript code refactor

### PR DESCRIPTION
**Pull request by trim21**
based on https://github.com/minio/minio-js/pull/1120 https://github.com/minio/minio-js/pull/1119

and close https://github.com/minio/minio-js/issues/1106

please don't review the details of this PR, it for staging the migration I have made.
--

all files have aleady been re-written in TS (with minor `any`)

just wait previous PR to me merged, the we can start to split this PR into smaller patch to seprated PR for code review.

We should migrate base files first, ts file should not import js file.

and minio.js (the `Client` class) is very large so we should create a typed base class and migrate methods progressively

### CODE REVIEW

File: src/main/client.mts [hunk 1]
```diff
@@ -0,0 +1,1655 @@
...
+  makeBucket(bucketName: string, region: Region, callback: NoResultCallback): void
+  makeBucket(bucketName: string, callback: NoResultCallback): void
+
+  // there is also a deprecated Backward Compatibility sign
+  // makeBucket(bucketName: string, makeOpts: MakeBucketOpt, callback: NoResultCallback): void
+  makeBucket(
+    bucketName: string,
+    arg2?: string | NoResultCallback,
+    arg3?: MakeBucketOpt | NoResultCallback,
+    arg4?: NoResultCallback
```
Comment (aldy505) [inline src/main/client.mts:839]: these parameters should be renamed to correct identifiers.

File: src/main/errors.mts [hunk 2]
```diff
@@ -14,122 +14,74 @@
...
 
-import ExtendableError from 'es6-error'
+class ExtendableError extends Error {
+  // es6 doesn't support new error cause
+  // and nodejs runtime will add stack automatically, no need to add it.
+  constructor(message?: string, opt?: ErrorOptions) {
+    super(message, opt)
+    this.name = this.constructor.name
+  }
+}
```
Comment (aldy505) [inline src/main/errors.mts:24]: 👍🏻

File: src/main/type.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
+import type { BinaryLike } from 'crypto'
+import type { IncomingMessage } from 'http'
+import type { Readable as ReadableStream } from 'stream'
```
Comment (aldy505) [inline src/main/type.ts:3]: shouldnt we use `... from "node:xxx"` for consistency?

File: src/main/type.ts [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+export type TagList = Record<string, string>
+export type EmptyObject = Record<string, never>
+export type VersionIdentificator = Pick<RetentionOptions, 'versionId'>
+export type Lifecycle = LifecycleConfig | null | ''
+export type Lock = LockConfig | EmptyObject
+export type Encryption = EncryptionConfig | EmptyObject
+export type Retention = RetentionOptions | EmptyObject
+export type IsoDate = string
+
+export interface GetObjectOpt {
```
Comment (aldy505) [inline src/main/type.ts:43]: I prefer using "GetObjectOptions" for naming things. 

but this should work fine if not directly exported to users.

**Comment by trim21** [comment]
@aldy505 I think it would be better to review the seprated pr instead this big one

File: src/main/type.ts (hunk 3 above)
Comment (trim21) [reply src/main/type.ts:3]: I personally prefer the node protocol, but I'm not sure if it should be enforced

File: src/main/s3-endpoints.mts [hunk 5]
```diff
@@ -14,10 +14,12 @@
  * limitations under the License.
  */
 
-import { isString } from './helpers.js'
+import { isString } from './asserts.mts'
+
+export const DEFAULT_REGION = 'us-east-1'
```
[the remark this answers is not in the document]
Comment (trim21) [reply src/main/s3-endpoints.mts:19]: this is a false alert I think

File: package.json [hunk 6]
```diff
@@ -67,7 +67,7 @@
     "through2": "^3.0.1",
     "web-encoding": "^1.1.5",
     "xml": "^1.0.0",
-    "xml2js": "^0.5.0"
+    "xml2js": "^0.4.15"
```
Comment (lavrov-lavrov) [inline package.json]: because of this edit i have this vulnerability: https://github.com/advisories/GHSA-776f-qx25-q3cc

Comment (trim21) [reply package.json]: that's a old commit

File: src/AssumeRoleProvider.ts [hunk 7]
```diff
@@ -1,13 +1,53 @@
-import Http from 'http'
-import Https from 'https'
+import type http from 'node:http'
```
Comment (aldy505) [inline src/AssumeRoleProvider.ts:1]: I believe we should put the LICENSE boilerplate at the top of the file

File: src/CredentialProvider.ts [hunk 8]
```diff
@@ -1,49 +1,50 @@
-import Credentials from './Credentials'
+import { Credentials } from './Credentials.ts'
```
Comment (aldy505) [inline src/CredentialProvider.ts:1]: License boilerplate here too

File: src/AssumeRoleProvider.ts (hunk 7 above)
Comment (trim21) [reply src/AssumeRoleProvider.ts:1]: Do not review this pr please, review seprated amall ones

Comment (trim21) [reply src/AssumeRoleProvider.ts:1]: https://github.com/minio/minio-js/pull/1119

Comment (aldy505) [reply src/AssumeRoleProvider.ts:1]: Oh do you want to use this PR as a base for migrating everything else into smaller chunks?

Comment (trim21) [reply src/AssumeRoleProvider.ts:1]: > Oh do you want to use this PR as a base for migrating everything else into smaller chunks?

Yes, so there is no need to review this PR.

Comment (trim21) [reply src/AssumeRoleProvider.ts:1]: I didn't expect this pr to me be merged.

Comment (aldy505) [reply src/AssumeRoleProvider.ts:1]: Alright. You should rename the PR then.

File: src/internal/type.ts [hunk 9]
```diff
@@ -56,3 +62,217 @@ export interface IRequest {
...
+export type ResultCallback<T> = (error: unknown | null, result: T) => void
+export type TagList = Record<string, string>
+export type EmptyObject = Record<string, never>
+export type VersionIdentification = { versionId?: string }
+export type Lifecycle = LifecycleConfig | null | ''
+export type Lock = LockConfig | EmptyObject
+export type Retention = RetentionOptions | EmptyObject
+export type IsoDate = string
+export type GetObjectOpt = {
+  versionId?: string
```
Comment (fflorent) [inline src/internal/type.ts:84]: I am depending on small changes of this PR, like the introduction of GetObjectOpt and StatObjectOpts.

Can I open a PR in order to introduce them without waiting the end of the refactoration or have I to wait? (or is there anything else I can do?)

Comment (trim21) [reply src/internal/type.ts:84]: The refactoration is already done, we just need to split this PR into small pieces for maintainer to review them. You can cherry-pick the part you need and create a seprated PR.

**Pull request closed without merging**